Track reasoning duration for models with hidden reasoning

Wolfsblvt opened the pull request:
Main goal was to enable reasoning/thinking time box for models that do not support sending back a reasoning string, but are know to actual do reasoning in the background.
I do like the idea of actually seeing the difference between thinking time and the actual full gen time for the message.

![image](https://github.com/user-attachments/assets/855b564d-017f-4246-ac8b-6003e64109a7)

### Changes
- Refactor reasoning handling from `StreamingProcessor` to `ReasoningHandler` for more logical structure
- Calculate reasoning time for some models that do reasoning, but do not send the reasoning string. (o1, o3, gemini experimental)
  - "Thinking..." and thought time will be shown during generation, but not be visible in chat history
  - To always show it, a small Custom CSS is needed
- Use reasoning handling state for processing and UI for easier CSS selectors and logic
- Add HTML attributes for reasoning state and reason duration
- Streamline internal function to redraw reasoning UI to always use the same functionality (DRY)
- Change reasoning start time to match the `gen_time` property generated messages
- Small UI changes on reasoning display for non-reasoning-showing models
- Remove Placeholder no-width-space string being persisted into the actual message object

### Fixes
- Make sure `STREAM_REASONING_DONE` only fires once and at the correct moment when reasoning streaming has finished
- Improve and fix several states of editing, collapsed and not collapsed states of messages and reasoning

### General Changes
- Add global `data-generating="true"` attribute to `body` to signify an ongoing generation
  - Switch most of the existing js/jQuery code to hide/show message elements during generation to CSS selectors with the new data attribute

<!-- Put X in the box below to confirm -->

## Checklist:

- [x] I have read the [Contributing guidelines](https://www.youtube.com/watch?v=vDWlhQjGjoc).


## Copilot
This pull request focuses on refactoring the handling of reasoning in the `public/script.js` file. The primary change is the introduction of a new `ReasoningHandler` class to manage reasoning-related functionality, which simplifies the code and improves maintainability.

Key changes include:

### Refactoring Reasoning Handling:

* Replaced direct reasoning handling with the `ReasoningHandler` class in `public/script.js` [[1]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L272-R272) [[2]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L3150-R3142) [[3]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L3439-R3376).
* Removed `reasoning` and `reasoningDuration` parameters from various functions and replaced them with calls to `updateReasoningUI` [[1]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L2226-L2227) [[2]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L2252-R2257) [[3]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L2458-R2457).
* Updated `formatGenerationTimer` function to include reasoning duration [[1]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R2577-R2584) [[2]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R2598).

### Code Cleanup and Simplification:

* Removed unnecessary reasoning-related DOM elements and methods from `StreamingProcessor` class [[1]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R3118-L3134) [[2]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L3181-L3193).
* Consolidated reasoning updates into the `ReasoningHandler` class, reducing redundancy [[1]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L3242-L3282) [[2]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L3308-R3255).

### Additional Updates:

* Updated imports in `public/scripts/reasoning.js` to include `messageFormatting`.
* Ensured reasoning UI is reset appropriately in the `swipe_right` function.

Wolfsblvt committed:
Add thinking time for hidden reasoning models

- Streamline reasoning UI update functionality
- Add helper function to identify hidden reasoning models
- Fix/update reasoning time calculation to actually utilize start gen time
- Fix reasoning UI update on swipe
- add CSS class for hidden reasoning blocks (to make it possible to hide for users)

Wolfsblvt committed:
Add gemini pro to hidden thinking models

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -4995,6 +4995,54 @@ export function isImageInliningSupported() {
+ * Check if the model supports reasoning, but does not send back the reasoning
+ * @returns {boolean} True if the model supports reasoning
+ */
+export function isHiddenReasoningModel() {
```
Maybe move to reasoning.js?

Cohee1207 committed:
Move isHiddenReasoningModel

Wolfsblvt committed:
Add thinking time to gen timer tooltip

Wolfsblvt committed:
Refactoring StreamProcessor -> ReasoningHandler

Wolfsblvt committed:
More coding on ReasoningHandler, fixing stuff

Wolfsblvt committed:
Update reasoning CSS hide hidden reasoning again

Wolfsblvt committed:
Small improvements to reasoning UI

Wolfsblvt committed:
Merge branch 'staging' into hidden-reasoning-tracking

Wolfsblvt committed:
Update hidden reasoning model function

- Now more flexibly supports different API providers, and uses direct model names
- Fixes hidden reasoning time for OpenRouter and custom endpoints

Wolfsblvt committed:
More fixes on reasoning

- Fix resetting reasoning on swipes
- Fix not updating reasoning time/end on gen when trim spaces was enabled
- Fix hidden model checking not working

Cohee1207 commented on public/scripts/reasoning.js:
```diff
@@ -71,21 +71,346 @@ export function extractReasoningFromData(data) {
+        return false;
+    }
+
+    switch (oai_settings.chat_completion_source) {
```
Let's export a function `getChatCompletionModel` from openai.js and reuse it here. It's very easy to miss when adding a new CC source otherwise.

Wolfsblvt replied:
I didn't know about that one even. Only knew about `getModelOptions` from slash commands, but I really didn't want to dip into slash commands, or abuse the UI elements here.
This should be much easier.

Edit: Done.

Cohee1207 committed:
Gemini: don't send "thinkingConfig" to backend
As it does absolutely nothing

Wolfsblvt committed:
Fix and shorten isHiddenReasoningModel

Cohee1207 committed:
Merge branch 'staging' into hidden-reasoning-tracking

Cohee1207 committed:
Fix empty lines in generation timer

Wolfsblvt committed:
Make "Thinking..." default to fix weird flicker

Wolfsblvt committed:
Attempt to fix wrong display behavior

Wolfsblvt committed:
Yoink reasoning placeholder from being saved

Wolfsblvt committed:
Rework reasoning CSS selectors and fixes

- Remove reliance on :empty state of reasoning textbox, now using the actual css class on mes (".mes.reasoning")
- Fix reasoning add not working when Auto-Expand is not enabled
- Clean setting of dataset attributes, now removing when null

Wolfsblvt committed:
Allow editing/adding hidden reasoning blocks

- Fix cancel not cancelling correctly
- Allow editing/adding of hidden reasoning blocks
- Make style and tooltip to show on edit that hidden reasoning is present
- Update reasoning UI on edit cancel - to be sure

Cohee1207 committed:
Merge branch 'staging' into hidden-reasoning-tracking

Cohee1207 reviewed (approved):
LGTM

Wolfsblvt committed:
Swiping doesn't need to reset reasoning UI

Wolfsblvt committed:
Add data attribute during generation

- Fix reasoning buttons still appearing
- Allow easier customization of HTML elements via CSS that appear during gen

Wolfsblvt committed:
Show "Thinking..." while generation is happening

- Show thinking/thought for while generation is happening, even when hidden thinking
- Fix empty reasoning having toggle-able block

Wolfsblvt committed:
Switched even more to "generating" CSS

Wolfsblvt committed:
Re-add reasoning placeholder check to prompt build

- Was removed because placeholder is not needed anymore. But for compatibility with old saved chats from staging that might have this, we should keep it for a while (or forever)

Cohee1207 committed:
Merge branch 'staging' into hidden-reasoning-tracking

Wolfsblvt committed:
Add toggle for showing hidden reasoning
